### ISSUE TITLE
Script not compilable

### ISSUE THREAD

**Issue by phil294** [bobsayshilol/luajit-decomp#2]
The exe runs fine, however in AutoIt code: `Error: Subscript used on non-accessible variable` in line 830 (decoder_new.au3). I believe it is due to line 45 (_ArrayAdd returns error code 1). Having looked at the documentation briefly, I am amazed this actually every compiled - a 2D array is being constructed but it is declared as a 1D-array? But really, idk
I guess this would be a minor fix. Cheers

**Comment by bobsayshilol** [comment]
Wow my old code is abysmal. I'll be totally honest I don't remember much about AutoIt or how it works, but I think you might be running it on an invalid/incompatible LuaJIT file.

Lines 38-47 make the 1D array `$faa` into a 2D array where the first index is the function number and the elements in that sub-array are the lines that correspond to that functions instructions. It looks like the loop assumes the first line it's going to be reading starts with `-- B`, ie a function start, which means that if you're erroring on that line then the input file likely isn't LuaJIT.

For it to be crashing on line 830 `$func` must be invalid, which would make sense if `$faa` was never populated, and as you can see in `handleLine()` the function declaration lines should be written out verbatim to `$out` so it should appear when being read back in.

I'd suggest opening up `test.asm` and checking that the first line does have a `-- B` at the top, and logging `$l` on the first line of `handleLine()` to check that it's reading the correct file. If I had a guess I'd say that running it from an AutoIt editor doesn't use the working directory you might expect, so it wouldn't be able to either run the luajit command at the top of the script or read `test.asm` after processing it. I should probably have checked that the `RunWait()` call on line 4 actually returned success, but I'd rather rewrite this whole thing than do a bug fix to terrible old code :P

**Comment by phil294** [comment]
Thank you for your response and the time writing this up!
When I opened this issue three years ago, I did not have problems decompiling a lua file with your exe, I just could not reproduce the exe from the au3 source code. I am pretty certain I even tried various au3 versions... but I now tried to do it once again, e.g. with https://www.script-example.com/themen/AutoIT-Online-Compiler.php (v3.3.6.1, same as in your old exe file), and it all worked without problems. I don't know what I did wrong back then? Maybe confused 64/32 bit environment? Anyway, this issue is invalid, I'll close it.

I also had a look at your promising lua reimplementation, it looks great, very readable for a change! :-) By replacing `assert(func.globals[...])` with `assert(func.globals[...] or comment)` in various places, I was also able to get it running with some test scripts as well. But I am sure you left out comment parsing on purpose so far. But TBH I am still more than happy with your old decompiler. Thank you once again.
